vdom/core.py

```diff
@@ -8,19 +8,33 @@
 This module provides the core implementation for the VDOM (Virtual DOM).
 
 """
-def toJSON(el):
-    """Convert an element to JSON"""
-    if(type(el) is str):
-        return el
+
+from jsonschema import validate, Draft4Validator, ValidationError
+
+def toJSON(el, schema=None):
+    """Convert an element to JSON
+
+    If you wish to validate the JSON, pass in a schema via the schema keyword argument.
+    If a schema is provided, this raises a ValidationError if JSON does not match the schema.
+    """
     if(type(el) is list):
-        return list(map(toJSON, el))
-    if(hasattr(el, 'tagName') and hasattr(el, 'attributes')):
-        return {
+        json_el = list(map(toJSON, el))
+    elif(hasattr(el, 'tagName') and hasattr(el, 'attributes')):
+        json_el = {
             'tagName': el.tagName,
             'attributes': el.attributes,
             'children': toJSON(el.children)
         }
-    return el
+    else:
+        json_el = el
+
+    if schema:
+        try:
+            validate(instance=json_el, schema=schema, cls=Draft4Validator)
+        except ValidationError as e:
+            raise ValidationError("Your object didn't match the schema: {}. \n {}".format(schema, e))
```

Comment:
Since `toJSON` is recursive, we may want to move the recursive calls to `_toJSON` and only do the validation at the top.

Reply:
that sounds like a really good idea!

Reply:
OR, map to `schema = None` at all but the top level of those recursions

Reply:
except for the fact that python's map doesn't work that way…

Reply:
Ugh… three approaches that don't involve making `_toJSON` a thing, only one of them allows keyword arguments:

1. use `map` and create an iterable of `None`s
```python
map(toJSON, el, [None]*len(el))
```
2. use `itertools.starmap` and repack every top level element of `el` with a `None`
```python
itertools.starmap(toJson, [(e, None) for e in el])
```
3. use `functools.partial` to create a function that already has `None` baked in (ala [this SO](https://stackoverflow.com/questions/13499824/using-map-function-with-keyword-arguments))
```python
map(functools.partial(toJSON, schema=None), el)
```

of the three I like the last the most, otherwise we are effectively locking in the keyword argument as a positional argument. 

However, my gut reaction is that `functools.partial` is mildly beginner unfriendly; but I can do any of these solutions or just create a private `_toJSON` that effectively just is the `functools.partial` bit.

Reply:
I'm dumb… as stated in the other comment, we don't need to do this so long as schema defaults to `None` we were already not applying the schema recursively. 

We'd need to take this approach if we ever wanted to set the schema to anything _but_ `None`.